### DISCUSSION TITLE
Routing - Optional Parameters

### DISCUSSION THREAD

**Discussion by izadoramatias** [symfony/symfony#52904] [Q&A]
I am reading/studying symfony docs about routing, but I was confused about the use of "!" in a route optional parameter. The documentation says:

> If you want to always include some default value in the generated URL (for example to force the generation of /blog/1 instead of /blog in the previous example) add the ! character before the parameter name: /blog/{!page}

But, in my tests, if I add the "!" in the parameter, I got an _404 error, saying that the route was not found._ But in theory, if I want to force the use of a default value for a parameter, if one is not entered, I should use "!", right? That's what I understood from the documentation. I apologize for any misunderstandings

This is the base documentation: [Optional Parameters](https://symfony.com/doc/6.3/routing.html#optional-parameters)
This is the code I wrote to test the docs: 
```php
class RouteController
{
    #[Route('/route/{!page}', name: 'blog_list', requirements: ['page' => '\d+'])]
    public function list(int $page = 1): Response
    {
        return new Response();
    }
}
```

**Comment by MatTheCat** [comment]
Indeed the documentation does not mention this has not only an impact on generation, but also on matching (since https://github.com/symfony/symfony/pull/29770). Are you up for a PR..?

**Reply by izadoramatias** [reply]
Yeah, but what should I do exactly? I mean, I never did a PR before

**Reply by MatTheCat** [reply]
Nice! You can read https://symfony.com/doc/current/contributing/documentation/overview.html; this should be enough to get started.

**Reply by izadoramatias** [reply]
Great!!! Thank you so much.
But before I open a PR, I have a question. Maybe I dont undesrtand exactly the documentation. What  does the documentation snippet below mean? Can you give me a use case for this character?

> If you want to always include some default value in the generated URL (for example to force the generation of /blog/1 instead of /blog in the previous example) add the ! character before the parameter name: /blog/{!page}

**Reply by MatTheCat** [reply]
A routing parameter is optional as long as it is a trailing one and it has a default value. In that case its value can be omitted when generating or matching the URI, e.g. if I have a `blog` route whose path is `/blog/{page?1}`, then

- `$urlGenerator->generate('blog')` will return `/blog`
- both `/blog` and `/blog/1` URIs will match the `blog` route and `page` will be `1`

But sometimes you want a **required** trailing parameter with a default value (#29770 calls them **important**). Think about an `image` generation route like `/image.{_format?jpg}` for example: you probably don’t want the `/image.` URI to match it. By prefixing `_format` with `!`,

- `$urlGenerator->generate('image')` will return `/image.jpg`
- `/image.` won’t match the `image` route
- `/image.jpg` will match the `image` route and `_format` will be `jpg`

Hope I was able to clarify?
